### PR TITLE
Update notifications for core and apps

### PR DESCRIPTION
**Pull request by nickvergessen**
- [x] Trigger notifications for new updates in a background job
- [x] Parse the notifications
- [x] Make notification user group configurable in the UI
- [x] ~~Automatically delete the notification when an app/core was updated~~ => No hooks, needs to move to "on rendering"
- [x] Add unit tests ?!

### CODE REVIEW

**Comment by nickvergessen** [comment]
Done, ready for review

**Comment by nickvergessen** [comment]
@LukasReschke @VicDeo @rullzer @PVince81

File: apps/updatenotification/js/admin.js [hunk 1]
```diff
@@ -39,8 +39,16 @@ $(document).ready(function(){
...
 		});
 	});
+
 	$('#release-channel').change(function() {
 		var newChannel = $('#release-channel').find(":selected").val();
+
+		if (newChannel === 'git' || newChannel === 'daily') {
+			$('#oca_updatenotification_groups em').removeClass('hidden');
+		} else {
+			$('#oca_updatenotification_groups em').addClass('hidden');
```
Comment (PVince81) [inline apps/updatenotification/js/admin.js]: protip: `$(...).toggleClass('hidden', boolValue)`

Comment (nickvergessen) [inline apps/updatenotification/js/admin.js]: can't toggle, because then I need to catch the case when you change from stable to product and not toggle...

**Comment by PVince81** [comment]
Will the notification be recreated during the next update check or is there code that makes sure the notification only appears once ? I didn't seem to find it, maybe it's a feature of the notifications app ?

File: apps/updatenotification/js/admin.js (hunk 1 above)
Comment (nickvergessen) [inline apps/updatenotification/js/admin.js]: same for beta, and then it's messier than now

**Comment by nickvergessen** [comment]
Yes, each new update will generate a new notification, but before that we delete the old notifications, in case one had them still unread.
Also the app removes the notifications, when the update has already been installed, so the other admins don't need to click the notification away

File: apps/updatenotification/lib/Notification/BackgroundJob.php [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+	/**
+	 * @return string[]
+	 */
+	protected function getUsersToNotify() {
+		if ($this->users !== null) {
+			return $this->users;
+		}
+
+		$notifyGroups = json_decode($this->config->getAppValue('updatenotification', 'notify_groups', '["admin"]'), true);
```
Comment (PVince81) [inline apps/updatenotification/lib/Notification/BackgroundJob.php]: if no groups were selected, disable the background job completely (or make it no-op)
In some scenarios an admin might not want to have a bkg job pinging the app store for every app regularily.

Comment (PVince81) [inline apps/updatenotification/lib/Notification/BackgroundJob.php]: Ok, maybe one can just disable the update notification app

File: apps/updatenotification/js/admin.js (hunk 1 above)
Comment (PVince81) [inline apps/updatenotification/js/admin.js]: Not really, this code should behave the same as yours, but shorter:
`$('#oca_updatenotification_groups em', !(newChannel === 'git' || newChannel === 'daily'))`

Anyway, it's not a biggie.

**Comment by PVince81** [comment]
App update notification doesn't seem to work for me.
What I did:

1. Install the Notes app from Rekkyt from the app store
1. Edit apps/notes/info.xml and set the version to 2.0.0 (it was 2.0.1)
1. `update oc_jobs set last_run=null;`
1. run CLI cron
1. Refresh the web page

There are no new entries in `oc_notifications`.
If I go to the apps page, the yellow notification appears.

The core update notification works though, so the jobs does run but something might not work correctly with apps updates.

**Comment by PVince81** [comment]
Just debugged a bit, and I saw it go here: https://github.com/owncloud/core/pull/24444/files#diff-b720db44b64d429650b1fa2d2bbbc777R129 even though there is no entry in `oc_notifications` (was never for this app)

Ok, so I tried doing this: `update oc_appconfig set configvalue="2.0.0" where configkey="notes" and appid="updatenotification";`.

Now when I run cron, there's an entry in `oc_notifications`.
```
MariaDB [owncloud]> select * from oc_notifications;
+-----------------+--------------------+-------+------------+-------------+-----------+------------------+--------------------+---------+--------------------+----------------------------------------------------+---------+
| notification_id | app                | user  | timestamp  | object_type | object_id | subject          | subject_parameters | message | message_parameters | link                                               | actions |
+-----------------+--------------------+-------+------------+-------------+-----------+------------------+--------------------+---------+--------------------+----------------------------------------------------+---------+
|               3 | updatenotification | admin | 1462890969 | notes       | 2.0.1     | update_available | []                 |         | []                 | http://localhost/index.php/settings/apps#app-notes | []      |
+-----------------+--------------------+-------+------------+-------------+-----------+------------------+--------------------+---------+--------------------+----------------------------------------------------+---------+
```

However when I refresh the web UI, that entry just disappears and nothing appears in the UI !

**Comment by nickvergessen** [comment]
I guess because your cheat was incomplete and the installed_version in appconfig is 2.0.1?

**Comment by PVince81** [comment]
Indeed. Works now :+1:

**Comment by nickvergessen** [comment]
Second review @schiesbn @ChristophWurst @rullzer

**Comment by rullzer** [comment]
:+1:

**Comment by PVince81** [comment]
@nickvergessen mind rebasing so we can see at least the basic DB tests pass ? Thanks (results timed out)

**Comment by nickvergessen** [comment]
I just ran them locally and they all passed, but sure can slap jenkins in a bit

**Comment by PVince81** [comment]
@nickvergessen here, use this :boot:

**Comment by nickvergessen** [comment]
Done

**Comment by PVince81** [comment]
CC @carlaschroder @jospoortvliet for this enhancement

**Comment by nickvergessen** [comment]
Docs in https://github.com/owncloud/documentation/issues/2449

**Pull request merged**